examples/vision/adamatch.py

```diff
@@ -0,0 +1,606 @@
+target_ds_s = create_individual_ds(svhn_train, strong_augment, source=False)
+final_target_ds = tf.data.Dataset.zip((target_ds_w, target_ds_s))
+
+"""
+Here's how a single image batche looks like:
+
+![](https://i.imgur.com/aver8cG.png)
+"""
+
+"""
+## Utilities for loss calculation utilities and weight scheduler
+"""
+
+
+def compute_loss_source(source_labels, logits_source_w, logits_source_s):
+    loss_func = keras.losses.CategoricalCrossentropy(from_logits=True)
+    # First compute the losses between original source labels and
+    # predictions made on the weakly and strongly augmented versions
+    # of the same images.
+    w_loss = loss_func(source_labels, logits_source_w)
+    s_loss = loss_func(source_labels, logits_source_s)
+    return w_loss + s_loss
+
+
+def compute_loss_target(target_pseudo_labels_w, logits_target_s, mask):
+    loss_func = keras.losses.CategoricalCrossentropy(from_logits=True, reduction="none")
+    target_pseudo_labels_w = tf.stop_gradient(target_pseudo_labels_w)
+    # For calculating loss for the target samples, we treat the pseudo labels
+    # as the ground-truth. These are not considered during backpropagation
+    # which is a standard SSL practice.
+    target_loss = loss_func(target_pseudo_labels_w, logits_target_s)
+
+    # More on `mask` later.
+    mask = tf.cast(mask, target_loss.dtype)
+    target_loss *= mask
+    return tf.reduce_mean(target_loss, 0)
+
+
+"""
+## Subclassed model for AdaMatch trainer
+
+The figure below presents the overall workflow of AdaMatch (taken from the
+[original paper](https://arxiv.org/abs/2106.04732)):
+
+![](https://i.imgur.com/1QsEm2M.png)
+
+Here's a brief step-by-step breakdown of the workflow:
+
+1. We first retrieve the weakly and stronhly augmented pairs of images from source and
+target images.
+2. Then we prepare two concatenated copies:
+
+    i. One where all the two pairs are concatenated.
+
+    ii. One where only the source image pair is concatenated.
+3. We make two forwarded passes through the underlying learning model:
+
+    i. The first forward pass encompasses the concatenated copy obtained from **2.i**. In
+this forward pass, the [Batch Normalization](https://arxiv.org/abs/1502.03167) statistics
+are updated.
+
+    ii. In this forward pass, we only use the concatenated copy obtained from **2.ii**.
+    Also, in this case, Batch Normalization layers are run in inference mode.
+4. Then the respective logits are computed for both the forward passes.
+5. The logits then go through a series of transformations introduced in the paper (which
+are discussed shortly).
+6. We then compute the losses and update the gradients of the underlying model.
+"""
+
+
+class AdaMatch(keras.Model):
+    def __init__(self, model, tau=0.9):
+        super(AdaMatch, self).__init__()
+        self.model = model
+        self.tau = tau  # Denotes the confidence threshold
+        self.loss_tracker = tf.keras.metrics.Mean(name="loss")
+        self.total_steps = TOTAL_STEPS
```

Comment:
Make this a constructor argument, then do `model = AdaMatch(total_steps= TOTAL_STEPS)`. This improves code reusability. The internal implementation of a model should not depend on outer global constants.